examples/hosts/app-integration/schema-upgrade/src/migrator.ts

```diff
@@ -0,0 +1,174 @@
+            throw new Error("Cannot perform migration, we are currently trying to load");
+        }
+
+        const migratable = this._currentMigratable;
+        const acceptedVersion = migratable.acceptedVersion;
+        if (acceptedVersion === undefined) {
+            throw new Error("Expect an accepted version before migration starts");
+        }
+
+        const doTheMigration = async () => {
+            // It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.  Currently
+            // this fails the migration gracefully and emits an event so the app developer can know they're stuck.
+            // Ideally the app developer would find a way to acquire a new ModelLoader and move forward, or at least
+            // advise the end user to refresh the page or something.
+            // TODO: Does the app developer have everything they need to dispose gracefully when recovering with a new
+            // ModelLoader?
+            const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
+            if (!migrationSupported) {
+                this.emit("migrationNotSupported", acceptedVersion);
+                this._migrationP = undefined;
+                return;
+            }
+
+            const extractedData = await migratable.exportData();
+
+            // If we needed to transform the extracted data, we would do it here.  In this demo, the export/import
+            // format is unchanged between the two versions, so there's no transformation needed.
+            // TODO: think about how we might enable a transform while allowing Migrator to remain generic?
+            // E.g. have some callback with from/to version that does the transform?
+
+            const createResponse = await this.modelLoader.createDetached(acceptedVersion);
+            const migratedModel: IMigratableModel = createResponse.model;
+            if (!migratedModel.supportsDataFormat(extractedData)) {
+                throw new Error("New model doesn't support extracted data format");
+            }
+            // TODO: Probably should also validate at proposal time that the model we're proposing will be
+            // able to import the exported format of the current container (taking into consideration our format
+            // transform options).  Not all clients might agree about support if some have old ModelLoaders -- these
+            // clients could use this opportunity to dispose early and try to get new ModelLoaders.
+            await migratedModel.importData(extractedData);
+
+            // Before attaching, let's check to make sure no one else has already done the migration
+            // To avoid creating unnecessary extra containers.
+            if (migratable.getMigrationState() === MigrationState.migrated) {
+                return;
+            }
+
+            // TODO: Support retry
+            // TODO: Use TaskManager here to reduce container noise.
```

Comment:
Added a comment in latest - it's a little nuanced since I think we'll want to keep racing the local portion, and only negotiate for the attach/finalize.  Also I think some of this might change a bit once we add retry logic.